numpy/fft/_pocketfft_umath.cpp

```diff
@@ -20,41 +20,47 @@
 
 #include "npy_config.h"
 
-#include "pocketfft/pocketfft.h"
+#include "pocketfft/pocketfft_hdronly.h"
+using pocketfft::detail::FORWARD;
+using pocketfft::detail::BACKWARD;
+using pocketfft::detail::get_plan;
+using pocketfft::detail::pocketfft_r;
+using pocketfft::detail::pocketfft_c;
+using pocketfft::detail::cmplx;
 
 /*
  * Copy all nin elements of input to the first nin of the output,
  * and any set any remaining nout-nin output elements to 0
  * (if nout < nin, copy only nout).
  */
-
+template <typename T>
 static inline void
 copy_data(char* in, npy_intp step_in, npy_intp nin,
-          char* out, npy_intp step_out, npy_intp nout, npy_intp elsize)
+          char* out, npy_intp step_out, npy_intp nout)
```

Comment:
Two comments:
1. Let's use `T *out` and `T *in` here and then follow up with using `*out = *in` below and also `*out = 0`.  (I.e. remove all memcpy/memset, it was just annoying without templating.) 
2. Don't use that `pocketfft:detail::cmpl` thingy, just use `std::complex` (with `#include <complex>`.  However, I am not sure if there be dragons with migw longdouble:  You may have to use the LongDouble defined in `_core/src/npstd.hpp` (that pulls in complex also).

Reply:
> 1. Let's use `T *out` and `T *in` here and then follow up with using `*out = *in` below and also `*out = 0`.  (I.e. remove all memcpy/memset, it was just annoying without templating.)

And adjust the step sizes accordingly in the call, I assume?

Reply:
> 2\. Don't use that `pocketfft:detail::cmpl` thingy, just use `std::complex` (with `#include <complex>`.  However, I am not sure if there be dragons with migw longdouble:  You may have to use the LongDouble defined in `_core/src/npstd.hpp` (that pulls in complex also).

I think I need `cmplx` to be able to pass it in to the pocketfft functions without getting warnings (this likely mostly because I'm using internal routines).

Reply:
Ah, sorry, I forgot about striding...

You can't do that, since step sizes might be half-way or so, we could assign via `*(T *)out = 0`, which is still a bit nicer I would say?

Reply:
OK, I mainly thought to replace it to not use it too much, and I thought that assignments seem to maybe not work for that one, but maybe they do anyway...

If this seems no improvement, then don't worry too much, that seemed to me like it might with a brief look, but not sure.

Reply:
@seberg - Do you mean something like,
```
        if (step_in == 1 && step_out == 1)) {
            for (npy_intp i = 0; i < ncopy; i++) {
                out[i] = in[i];
            }
        }
        else {
            T *ip = in, *op = out;
            for (npy_intp i = 0; i < ncopy; i++, ip += step_in, op += step_out) {
                *op = *ip;  // Or without pointers? out[op] = in[ip]
            }
        }
```
Or just a single loop? (but then it is less obvious whether the compiler can still optimize the contiguous case)

Reply:
I tried and liked it at some level, but got stuck at being able to set `out[i] = 0` which `cmplx` type does not know how to interpret. So maybe indeed best to just keep things as they are...

Reply:
Right, I thought that would the case and that using `std::complex` would allow it.  Have to look at the rest, but for now let's just keep it as is, I guess.